### PR TITLE
Move cmake_[windows|linux]_gpu to CUDA-11.3

### PR DESCRIPTION
**Pull request by malfet**
Also, made a small change to CUDA install script on windows:
`echo %errorlevel%` would always return 0, and save cuda install log into the home folder rather than in current directory

cc @seemethere

### CODE REVIEW

**Comment by datumbox** [comment]
@malfet I don't think that works and neither passing `CU_VERSION`. I gave it a try at #4921 and it seems the fix will be more elaborate as it requires installing missing libs, changing the image type etc. I closed it for now and plan to revisit it. If you have pointers let me know. Thanks!

File: packaging/windows/internal/cuda_install.bat [hunk 1]
```diff
@@ -222,9 +224,9 @@ set "NVTOOLSEXT_PATH=%ProgramFiles%\NVIDIA Corporation\NvToolsExt\bin\x64"
 if not exist "%ProgramFiles%\NVIDIA GPU Computing Toolkit\CUDA\v%CUDA_VERSION_STR%\bin\nvcc.exe" (
```
Comment (mszhanyi) [inline packaging/windows/internal/cuda_install.bat:224]: Shall we move this section to line 195

**Comment by mszhanyi** [comment]
> Also, made a small change to CUDA install script on windows: `echo %errorlevel%` would always return 0, and save cuda install log into the home folder rather than in current directory

@malfet , errorlevel would be non-zero if setup.exe failed.
https://app.circleci.com/pipelines/github/pytorch/pytorch/372302/workflows/8a242e28-ad44-4af6-a53a-a031cb9e92d2/jobs/15706071

```
C:\w\b\windows\temp_build\cuda>start /wait setup.exe -s thrust_11.3 nvcc_11.3 cuobjdump_11.3 nvprune_11.3 nvprof_11.3 cupti_11.3 cublas_11.3 cublas_dev_11.3 cudart_11.3 cufft_11.3 cufft_dev_11.3 curand_11.3 curand_dev_11.3 cusolver_11.3 cusolver_dev_11.3 cusparse_11.3 cusparse_dev_11.3 npp_11.3 npp_dev_11.3 nvrtc_11.3 nvrtc_dev_11.3 nvml_dev_11.3 -loglevel:6 -log:"C:\w\b\windows\temp_build\cuda/cuda_install_logs" 

C:\w\b\windows\temp_build\cuda>echo -469237760 
-469237760
```
in https://circleci.com/api/v1.1/project/github/pytorch/pytorch/15706071/output/104/0?file=true&allocation-id=6129e01c942d271d4e1ef010-0-build%2F7371EF63

File: packaging/windows/internal/cuda_install.bat (hunk 1 above)
Comment (malfet) [reply packaging/windows/internal/cuda_install.bat:224]: Sounds like a good idea. Let me do it as part of separate PR

**Comment by malfet** [comment]
> @malfet , errorlevel would be non-zero if setup.exe failed. 
Yes, but `errorlevel` is not the same as `%errorlevel%`, and as far as I understand, general advise for batch scripts is to use `if errorlevel`, rather than `if %errorlevel%`, see https://ss64.com/nt/errorlevel.html

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 3.12)
 project(torchvision)
-set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD 17)
```
Comment (datumbox) [inline CMakeLists.txt:3]: Is this necessary? My understanding is that TorchVision aligns with PyTorch
s approach and sets `CMAKE_CXX_STANDARD` to 14.

File: packaging/build_cmake.sh [hunk 3]
```diff
@@ -42,7 +42,14 @@ else
   PYTORCH_MUTEX_CONSTRAINT=''
 fi
 
-conda install -yq \pytorch=$PYTORCH_VERSION $CONDA_CUDATOOLKIT_CONSTRAINT $PYTORCH_MUTEX_CONSTRAINT $MKL_CONSTRAINT -c "pytorch-${UPLOAD_CHANNEL}"
+if [[ "$CU_VERSION" == "cu111" ]]; then
+  # cuda-toolkit=11.1 is only available on conda-forge
+  CONDA_EXTRA_FLAGS="-c conda-forge"
```
Comment (datumbox) [inline packaging/build_cmake.sh:47]: @fmassa I remember there was a discussion prior my leave on June about using forge, but not sure what was decided?

File: .circleci/regenerate.py [hunk 4]
```diff
@@ -268,9 +268,9 @@ def cmake_workflows(indentation=6):
         for device in device_types:
             job = {"name": f"cmake_{os_type}_{device}", "python_version": python_version}
 
-            job["cu_version"] = "cu102" if device == "gpu" else "cpu"
+            job["cu_version"] = "cu111" if device == "gpu" else "cpu"
```
Comment (datumbox) [inline .circleci/regenerate.py:271]: Do we have to align with PyTorch and use CUDA 11.3 for which we get nightly binaries OR it's OK to stick with 11.1? Are there any incompatibilities we should be aware of?

**Review by datumbox — COMMENTED** [review]
@malfet Thanks for your help in fixing this. I've added a couple of questions, let me know your thoughts.

File: CMakeLists.txt (hunk 2 above)
Comment (malfet) [reply CMakeLists.txt:3]: This is necessary to fix Windows compilation errors, where CUDA-11.x (which supports C++17 standard) could not be compiled using C++14 standard.

File: packaging/build_cmake.sh (hunk 3 above)
Comment (malfet) [reply packaging/build_cmake.sh:47]: conda-forge is still needed if one wants to install CUDA-11.1, but I will probably move the config to CUDA-11.3 and be done with it.

File: CMakeLists.txt (hunk 2 above)
Comment (datumbox) [reply CMakeLists.txt:3]: Thanks for the clarification. Makes sense. So I guess PyTorch Core will be moving as well to C++17, right?

File: .circleci/regenerate.py (hunk 4 above)
Comment (malfet) [reply .circleci/regenerate.py]: Yes, moving it to 11.3, don't want to mess us with 11.1 here

**Comment by malfet** [comment]
Ok, splitting this PR into 3: 
https://github.com/pytorch/vision/pull/4944 <- cherry-picks changes from audio back to vision that fixes CMake builds on Linux
https://github.com/pytorch/vision/pull/4945 <- Fix CUDA-11 Windows CMake builds

**Review by datumbox — APPROVED** [review]
LGTM, thanks!

**Pull request merged**